src/main/java/org/eclipse/pass/schema/SchemaInstance.java

```diff
@@ -1,39 +1,174 @@
+        if (orderedDeps.get(referencer) != null) {
+            if (orderedDeps.get(referencer).contains(schema)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
     /**
      * Counts the number of form properties in the schema
      *
      * @return int number of properties
      */
     private int countFormProperties() {
-        // TODO: Not implemented yet
-        return 0;
+        int num_properties = 0;
+        JsonNode properties = schema.at("/definitions/form/properties");
+        Iterator<String> properties_iterator = properties.fieldNames();
+        while (properties_iterator.hasNext()) {
+            num_properties++;
+            properties_iterator.next();
+        }
+        return num_properties;
     }
 
     /**
-     * Finds dependencies of this schema on other schemas
+     * Finds references in this schema to other schemas. Find by going through $ref
+     * tags in the schema
+     *
+     * @return List<String> List of URIs to schemas referenced by this schema
+     *
+     */
+    private void findRefs(JsonNode node, String pointer) {
+        Iterator<String> it = node.fieldNames();
+        it.forEachRemaining(k -> {
+            JsonNode value = node.get(k);
+            if (value.isValueNode()) {
+                if (k.equals(keyRef)) { // only do this if url is not null?
+                    refs.put(pointer.concat("/").concat(k), value.asText());
+                }
+            } else if (value.isObject())
+                findRefs(value, pointer.concat("/").concat(k));
+        });
+    }
+
+    /**
+     * Finds dependencies of this schema on other schemas. Find by going through
+     * $ref tags in the schema and adding only those that point to different schemas
      *
      * @return List<String> List of URIs to schemas depended on by this schema
+     *
      */
-    private List<String> findDeps() {
-        // TODO: Not implemented yet
-        return null;
+    private void findDeps() {
+        refs.forEach((path, ref) -> {
+            String schema = ref.split("#")[0];
+            if (!schema.equals("")) // only add refs to other schemas; not to itself
+                deps.put(path, ref);
+        });
+        orderedDeps.put(schema_name, deps.values());
+    }
+
+    public void dereference() {
+        deps.forEach((path, dep) -> {
+            path = path.substring(0, path.lastIndexOf("/"));
+            if (dep.charAt(0) == '#') {
+                resolveRef(path, dep.split("#")[1], schema);
+            } else { // referencing other schema
+                try {
+                    SchemaInstance ext_schema = SchemaFetcher.getSchema(dep.split("#")[0]);
+                    ((ObjectNode) schema).replace(path.split("/")[1],
+                            resolveRef(path, dep.split("#")[1], ext_schema.schema));
+                } catch (IOException e) {
+                    // TODO Auto-generated catch block
+                    e.printStackTrace();
```

Comment:
Should not swallow this exception. It should either be thrown (possibly wrapped in another exception) or logged.